### ISSUE TITLE
Collaboration with salem package

### ISSUE THREAD

**Issue by djhoese** [geoxarray/geoxarray#3]
@fmaussion How did your salem package never come up in #2288? Do you see an overlap with the geoxarray package? What are the goals of salem and where do you see it going in the future? Should geoxarray's proposed functionality be absorbed by xarray?

For reference for anyone else: https://salem.readthedocs.io/en/latest/index.html

Also @fmaussion we (pytroll group) or salem may be able to benefit from collaborating between salem and our pyresample and pycoast packages. Both packages probably need xarray accessors and better interfaces overall, but they work.
https://pyresample.readthedocs.io/en/latest/
https://pycoast.readthedocs.io/en/latest/

**Comment by fmaussion** [comment]
@djhoese thanks for reaching out.

salem was written at a time where much of the functionality I needed was not (yet) available, and I never intended to make a big thing out of it. I still think it can be useful to some (and it has a small user base), but I don't intend to develop it much further (mostly because I don't have time, but also because its scope is unclear, as you point out). When I wrote it I needed several things, which all have an equivalent in the python world but still have outstanding unsolved issues:

- graphics: nice, easy to do regional maps (as opposed to world maps) on map projections. For this, cartopy is much better, **but** for some reason yet unclear to me they [cannot label axes on lambert projs](https://stackoverflow.com/questions/27962953/cartopy-axis-label-workaround) and their proj objects are not (yet) compatible with pyproj (as you know). Also, salem is more high-level by hiding most of the matplotlib calls (and therefore less flexible than cartopy)
- simple GIS: map transforms, reprojections, region of interests based on shapes, etc. Here, pyresample or xESMF are much better **but** again, quite low-level in comparison. Also, salem knows only about a very small number of projections and does not have a CF proj parser. For these functionalities, I hope that geoxarray will replace salem entirely.
- read WRF files: here again, there is a much better competitor in terms of functionality [wrf-python](https://wrf-python.readthedocs.io), **but** (sorry for saying this) the API is quite <s>ugly</s> unpythonic because copy-pasted from the NCL language.

To answer your question: I don't think that salem needs to be included in any strategic discussion here.

The package will continue to be around because we use it in other (bigger) projects. I plan to maintain it but I won't develop it much further myself. I'm happy if some ideas are re-used by geoxarray, and I guess that if I find the time and energy I will replace some salem internals with geoxarray and cartopy tools once the issues mentioned above are solved.

**Comment by djhoese** [comment]
Thanks for the info @fmaussion. I'll keep this issue open for now until the features of salem are sincerely considered and a decision is made on whether they are out of scope for geoxarray, should be used from salem, if salem should be "adjusted" to have a more defined scope, or if you or I or someone else wants to create a package with a more defined scope for those features.

**Comment by fmaussion** [comment]
The way I see it (but I'm biased of course), what I would like geoxarray to do is mosly what I present in [salem's doc](https://salem.readthedocs.io/en/latest/xarray_acc.html), but in better:
- open any xarray dataset and understand its map proj
- reproject the data from one dataset to another (using pyresample under the hood)
- make a region of interest based on a shapefile or any geometry with a crs attached to it (like [regionmask](https://regionmask.readthedocs.io/en/stable/)?)
- provide entry points to cartopy for plotting 

So when geoxarray is out, I can do all my workflow by import three packages only: xarray, geoxarray, and cartopy. 

But maybe I'm not seeing this right (most of these functionalities are what drove the aborted [xmap](https://github.com/jhamman/xmap) project).

**Comment by fmaussion** [comment]
I'm going to write some of the things I've said above in the salem docs btw, so that prospective users know what to expect.

**Comment by hydrogeohc** [comment]
Is there any new update on this issue?

**Comment by djhoese** [comment]
@hydrogeohc do you mean with this specific issue or geoxarray as a project? I will make a new issue for the status of the project right now. As for this issue specifically, there is no collaboration planned between salem and geoxarray, or at least not more than geoxarray collaborating with any other project.

**Comment by hydrogeohc** [comment]
@djhoese  Thanks for your reply! I would like to have a latest update for your geoxarray project.

**Comment by djhoese** [comment]
Oops, meant to point out #6 which has my current thoughts.

**Comment by fmaussion** [comment]
Thanks @djhoese 

One thing which is independent from all the pangeo related grid discussions and that python is missing (AFAIK) is a NetCDF proj parser. This (I assume) what brought @hydrogeohc to geoxarray (xref: https://github.com/fmaussion/salem/issues/100 )

the fact that salem can parse *some* projs brings users to us, yet the huge majority of grids won't work of course.

I wonder how many other geospatial packages (e.g. MetPy) would be interested in having a unified way to parse geoloc information from NetCDF files.

**Comment by djhoese** [comment]
Yes this is very interesting. I've talked a little with @dopplershift about this in MetPy. MetPy has converters for some CF projections (not "most" iirc) for reading. The SatPy library (one I work on) has a couple that it supports when writing out CF files; @mraspaud added them.

I wonder if this should be put in its own package then and allow for WKT or PROJ strings to be returned or written out. I'm not sure what the status of the `cf-python` package is (https://cfpython.bitbucket.io/docs/latest/index.html) but maybe @dopplershift has some ideas.

**Comment by fmaussion** [comment]
> I wonder if this should be put in its own package then and allow for WKT or PROJ strings to be returned or written out.

I believe this is a very well defined use-case for a package. Being able to read many files (some of them non CF compliant) but write out CF-Compliant files only. "Reading" means roughly:
- understand the map projection of the data and give it in PROJ or WKT
- understand which are the coordinate variables and tell the user which is what

Of course I already omit many edge cases but this would be enough to make me (and salem) happy :wink:

**Comment by djhoese** [comment]
> Of course I already omit many edge cases but this would be enough to make me (and salem) happy 

Yes, I know metpy handles quite a few edge cases from what I remember the last time I read the source.

**Comment by dopplershift** [comment]
I'd be happy to contribute what we have somewhere and use that. (Preferrably without breaking our API, but I don't see that being a problem.)

The edge cases we currently handle:
* Stupid Geostationary projection using coordinates in radians rather than length units like **every other** CF projection; MetPy handles multiplying the radians by the satellite projection
* Conversion of all coordinates to meters for use directly with Cartopy
* Some...heuristics to figure out if we have, or should have, a lat/lon grid mapping

We also, under the metpy accessor, provide standard `x`, `y`, `vertical`, and `time` attributes to access the appropriate coordinate.

**Comment by dopplershift** [comment]
I'll also add that we've taken the tactic of adding projections as we find data using them, rather than exhaustively going through the CF document and adding everything; the latter approach would require synthesizing data for tests with no clue whether the synthetic data resembles real-world cases.

**Comment by rhugonnet** [comment]
Maybe relevant to this thread and for the way forward of `geoxarray`: we are planning to soon write an Xarray accessor for our packages `geoutils` and `xdem`, which aim to facilitate end-user analysis of rasters/vectors and DEMs, respectively. See more details at https://github.com/corteva/rioxarray/issues/687 and in our doc (https://geoutils.readthedocs.io/ and https://xdem.readthedocs.io/en/stable/). Right now all of our functionalities are based on `rasterio`, `geopandas` and `pyproj`. 

If I understand the aim of `geoxarray` correctly, and its relation to `salem` and `rioxarray`, there is absolutely no overlap and `geoutils` could eventually migrate towards relying directly on `geoxarray`? Or do you think there is need to coordinate on the writing of this Xarray accessor?

Regarding the accessor name, we could use `gu` to avoid overlap with `geo`, if that sounds good!

**Comment by djhoese** [comment]
I have very little experience with geopandas and really no experience with DEM data. I think geoxarray can be a dumping ground for anything that could be considered "generic" as far as geolocated data. I think a lot of the things mentioned in geoutils could be something that geoxarray could be depended on. Since I don't have much experience with DEMs I don't know if I can give a good opinion on whether geoxarray is the best place for processing data that has a vertical dimension to it, but I'd like it to be. Maybe that just means geoxarray isn't the best place for the functionality to go initially, just do the age of the project, but it could be put here in the future.

I would love (and plan to do some of these soon) for geoxarray to include functionality for coordinate transformation (between CRSes), coordinate generation (from semi-standard definitions to 1D/2D xarray coords), and boundary/polygon/intersection operations. There are also things like plotting conveniences that make some sense. I'd also like to have a generic resampling/reprojection interface that ties all these other libraries together. For example, `new_xarr_obj = old_xarr_obj.geo.resample(some_geo_definition, method="gdal[bilinear]")` or similar so that you don't have to worry about the interface differences between GDAL or pyresapmle or 1 of the other many resampling libraries and just let geoxarray figure it out.

**Comment by fmaussion** [comment]
hello! As far as salem is concerned, I see no major issue here. Salem is in "maintenance mode" since many years now (still being useful in several of my projects however).

**Comment by rhugonnet** [comment]
Thanks for the thoughts @djhoese! The relevance to vertical data depends on underlying support for 3D CRS transformation. Right now those aren't supported by `rasterio`, so we rely directly on `pyproj`. 3D CRSs have been evolving fast in the past years, it's getting more consistent.

From what you mention, your major choice seem to be if `geoxarray` will depend on a vector-centered package (`shapely`, `geopandas`?) or only focus on rasters.

In `geoutils`, we need to depend on both a raster-centered and vector-centered package to add wrappers on top for higher-level operations and analysis involving both. For more "generic" operations, our focus is simply on having a consistent API for rasters & vectors and how they are manipulated (same function names, arguments names, implicit metadata passing, consistent match-reference ops), but we let other packages figure out I/O or CRS definition/transformation details.

Our idea right now for the xarray accessor was to use `rioxarray` and `geocube` (to mirror the operations we do directly with `rasterio` and `geopandas`), but I need to think a bit more about it! As you say, can always move to, or add, `geoxarray` as a dependency later if it makes sense :slightly_smiling_face:

**Comment by djhoese** [comment]
Good points. I definitely have raster stuff in mind as my priority right now, but there are a lot of use cases on the horizon that vector/polygon operations are needed for. In pyresample we're starting to lean more and more on shapely for things like bounding box polygon operations (intersections, etc). I think pyproj and xarray as hard dependencies is the easy decision for geoxarray and I think shapely and/or geopandas is always an optional (but very useful) dependency.
